src/Workspaces/Core/Portable/Workspace/Solution/SolutionCompilationState.cs

```diff
@@ -0,0 +1,1317 @@
+                forkTracker: true);
+        }
+    }
+
+    /// <inheritdoc cref="SolutionState.WithHasAllInformation"/>
+    public SolutionCompilationState WithHasAllInformation(
+        ProjectId projectId, bool hasAllInformation)
+    {
+        return ForkProject(
+            this.Solution.WithHasAllInformation(projectId, hasAllInformation),
+            translate: null,
+            forkTracker: true);
+    }
+
+    /// <inheritdoc cref="SolutionState.WithRunAnalyzers"/>
+    public SolutionCompilationState WithRunAnalyzers(
+        ProjectId projectId, bool runAnalyzers)
+    {
+        return ForkProject(
+            this.Solution.WithRunAnalyzers(projectId, runAnalyzers),
+            translate: null,
+            forkTracker: true);
+    }
+
+    /// <inheritdoc cref="SolutionState.WithProjectDocumentsOrder"/>
+    public SolutionCompilationState WithProjectDocumentsOrder(
+        ProjectId projectId, ImmutableList<DocumentId> documentIds)
+    {
+        return ForkProject(
+            this.Solution.WithProjectDocumentsOrder(projectId, documentIds),
+            static stateChange => new CompilationAndGeneratorDriverTranslationAction.ReplaceAllSyntaxTreesAction(stateChange.NewProjectState, isParseOptionChange: false),
+            forkTracker: true);
+    }
+
+    /// <inheritdoc cref="SolutionState.AddProjectReferences"/>
+    public SolutionCompilationState AddProjectReferences(
+        ProjectId projectId, IReadOnlyCollection<ProjectReference> projectReferences)
+    {
+        return ForkProject(
+            this.Solution.AddProjectReferences(projectId, projectReferences),
+            translate: null,
+            forkTracker: true);
+    }
+
+    /// <inheritdoc cref="SolutionState.RemoveProjectReference"/>
+    public SolutionCompilationState RemoveProjectReference(StateChange stateChange)
+    {
+        return ForkProject(
+            stateChange,
+            translate: null,
+            forkTracker: true);
+    }
+
+    /// <inheritdoc cref="SolutionState.WithProjectReferences"/>
+    public SolutionCompilationState WithProjectReferences(
+        ProjectId projectId, IReadOnlyList<ProjectReference> projectReferences)
+    {
+        return ForkProject(
+            this.Solution.WithProjectReferences(projectId, projectReferences),
+            translate: null,
+            forkTracker: true);
+    }
+
+    /// <inheritdoc cref="SolutionState.AddMetadataReferences"/>
+    public SolutionCompilationState AddMetadataReferences(
+        ProjectId projectId, IReadOnlyCollection<MetadataReference> metadataReferences)
+    {
+        return ForkProject(
+            this.Solution.AddMetadataReferences(projectId, metadataReferences),
+            translate: null,
+            forkTracker: true);
+    }
+
+    /// <inheritdoc cref="SolutionState.RemoveMetadataReference"/>
+    public SolutionCompilationState RemoveMetadataReference(StateChange stateChange)
+    {
+        return ForkProject(
```

Comment:
This really should have a comment that this is different -- it looks so out of place compared to the rest.